### DISCUSSION TITLE
Connecting Crazyflie Bolt to offboard computer via UART

### DISCUSSION THREAD

**Discussion by tedzaremba** [bitcraze/discussions#1272] [Q&A] [answered]
Hi we're trying to pull data from the Crazyflie Bolt 1.1 to either a linux desktop via an FTDI cable or to our drone's UART port and the data doesn't seem to be populating. We're testing with a demo app made by @knmcguire ((https://github.com/knmcguire/crazyflie-firmware/tree/bolt-pos-through-uart)) and we have it compiling and flashed onto the bolt fine. We attached an oscilloscope to our Bolt and we can see that it's trying to transmit UART data, but we're not seeing it show up on either PuTTY or the python app. We've experimented with different baud rates, double checked the wiring, and deleted the debug statements.

The only difference between @knmcguire and our setup is that we're flashing over DFU while she's flashing via the CrazyRadio.

**Comment by knmcguire** [comment]
Hi! Glad you found the bitcraze discussions!

Just some questions such that we can debug this further:
- Which OS are you using? Is it detecting the ftdi cable?
- Are you using the python script provided with the example?
- Are you connecting ground of the bolt to the ftdi cable as well?
- What are the LEDs doing on the bolt? you should see 2 blue leds and one red led blinking at a constant rate.
- If you remove the loco positioning deck, what happens then? Does it work?

**Comment by tedzaremba** [comment]
Which OS are you using? Is it detecting the ftdi cable?
ubuntu 22.04

Are you using the python script provided with the example?
Yes 

Are you connecting ground of the bolt to the ftdi cable as well?
Yes

What are the LEDs doing on the bolt? you should see 2 blue leds and one red led blinking at a constant rate.
We only see the 2 blue LEDs lighting up.

If you remove the loco positioning deck, what happens then? Does it work?
The app seems to be running fine on the crazyflie-clients app  but we can't read data.

**Reply by knmcguire** [reply]
Ah the mystery continues! Here some more instructions

1. Could you share the content of the 'config' file in the build folder of the app? (to make sure that we have the same defines configured)
2. What do you see when you do `python3 -m serial.tools.list_ports -v` ?
3. Which port are you using? You can check if it is added to the right user group for rights with `ls -l /dev/ttyUSB...`

**Reply by tedzaremba** [reply]
Could you share the content of the 'config' file in the build folder of the app? (to make sure that we have the same defines configured)

<pre>
#
# Automatically generated file; DO NOT EDIT.
# Crazyflie Platform Configuration
#

#
# Build and debug options
#
CONFIG_CROSS_COMPILE="arm-none-eabi-"
# CONFIG_DEBUG is not set
# CONFIG_DEBUG_ENABLE_LED_MORSE is not set
# CONFIG_DEBUG_PRINT_ON_UART1 is not set
# CONFIG_DEBUG_DECK_IGNORE_OW is not set

#
# Platform configuration
#
# CONFIG_PLATFORM_CF2 is not set
CONFIG_PLATFORM_BOLT=y
# CONFIG_PLATFORM_TAG is not set
# CONFIG_PLATFORM_FLAPPER is not set
# CONFIG_BOLT11_BRUSHED is not set

#
# IMU configuration
#
# CONFIG_IMU_MADGWICK_QUATERNION is not set
CONFIG_IMU_MAHONY_QUATERNION=y

#
# Sensor configuration
#
# CONFIG_SENSORS_MPU9250_LPS25H is not set
CONFIG_SENSORS_BMI088_BMP388=y
# CONFIG_SENSORS_BOSCH is not set
# CONFIG_SENSORS_IGNORE_BAROMETER_FAIL is not set
CONFIG_SENSORS_BMI088_SPI=y
CONFIG_SENSORS_BMI088_I2C=y

#
# Power management configuration
#
# CONFIG_PM_AUTO_SHUTDOWN is not set

#
# Communication
#
# CONFIG_SYSLINK_RX_DMA is not set
CONFIG_ENABLE_CPX=y
CONFIG_ENABLE_CPX_ON_UART2=y
CONFIG_CPX_UART2_BAUDRATE=576000

#
# App layer configuration
#
# CONFIG_APP_API_TEST is not set
CONFIG_APP_ENABLE=y
CONFIG_APP_STACKSIZE=350
CONFIG_APP_PRIORITY=1

#
# Expansion deck configuration
#
CONFIG_DECK_FORCE="none"
CONFIG_DECK_ACTIVE_MARKER=y
CONFIG_DECK_AI=y
CONFIG_DECK_AI_WIFI_NO_SETUP=y
# CONFIG_DECK_AI_WIFI_SETUP_AP is not set
# CONFIG_DECK_AI_WIFI_SETUP_STA is not set
# CONFIG_DECK_BIGQUAD is not set
CONFIG_DECK_BUZZ=y
# CONFIG_DECK_CPPM is not set
# CONFIG_DECK_FLAPPER is not set
CONFIG_DECK_FLOW=y
# CONFIG_DECK_GTGPS is not set
CONFIG_DECK_LEDRING=y
CONFIG_DECK_LEDRING_DEFAULT_EFFECT=6
CONFIG_DECK_LEDRING_NBR_LEDS=12
CONFIG_DECK_LEDRING_DIMMER=0
CONFIG_DECK_LIGHTHOUSE=y
# CONFIG_DECK_LIGHTHOUSE_AS_GROUNDTRUTH is not set
CONFIG_DECK_LIGHTHOUSE_MAX_N_BS=4
CONFIG_DECK_LOCO=y
# CONFIG_DECK_LOCODECK_USE_ALT_PINS is not set
CONFIG_DECK_LOCO_NR_OF_ANCHORS=8
CONFIG_DECK_LOCO_ALGORITHM_AUTO=y
# CONFIG_DECK_LOCO_ALGORITHM_TWR is not set
# CONFIG_DECK_LOCO_ALGORITHM_TDOA2 is not set
# CONFIG_DECK_LOCO_ALGORITHM_TDOA3 is not set
# CONFIG_DECK_LOCO_FULL_TX_POWER is not set
# CONFIG_DECK_LOCO_TDOA3_HYBRID_MODE is not set
CONFIG_DECK_MULTIRANGER=y
CONFIG_DECK_OA=y
# CONFIG_DECK_SERVO is not set
CONFIG_DECK_USD=y
# CONFIG_DECK_USD_USE_ALT_PINS_AND_SPI is not set
CONFIG_DECK_ZRANGER=y
CONFIG_DECK_ZRANGER2=y
# CONFIG_DECK_CPX_HOST_ON_UART2 is not set
# CONFIG_DECK_ACS37800 is not set
# CONFIG_DECK_RPM is not set
# CONFIG_DECK_LOADCELL is not set

#
# Controllers and Estimators
#
CONFIG_CONTROLLER_AUTO_SELECT=y
# CONFIG_CONTROLLER_PID is not set
# CONFIG_CONTROLLER_INDI is not set
# CONFIG_CONTROLLER_MELLINGER is not set
# CONFIG_CONTROLLER_BRESCIANINI is not set
# CONFIG_CONTROLLER_OOT is not set
CONFIG_ESTIMATOR_KALMAN_ENABLE=y
# CONFIG_ESTIMATOR_KALMAN_TDOA_OUTLIERFILTER_FALLBACK is not set
# CONFIG_ESTIMATOR_UKF_ENABLE is not set
CONFIG_ESTIMATOR_OUTLIER_FILTERS=y
CONFIG_ESTIMATOR_AUTO_SELECT=y
# CONFIG_ESTIMATOR_KALMAN is not set
# CONFIG_ESTIMATOR_COMPLEMENTARY is not set
# CONFIG_ESTIMATOR_OOT is not set

#
# Motor configuration
#
# CONFIG_MOTORS_ESC_PROTOCOL_STANDARD_PWM is not set
CONFIG_MOTORS_ESC_PROTOCOL_ONESHOT125=y
# CONFIG_MOTORS_ESC_PROTOCOL_ONESHOT42 is not set
# CONFIG_MOTORS_ESC_PROTOCOL_DSHOT is not set
CONFIG_MOTORS_REQUIRE_ARMING=y
CONFIG_MOTORS_DEFAULT_IDLE_THRUST=0
CONFIG_MOTORS_DEFAULT_PROP_TEST_PWM_RATIO=0
CONFIG_MOTORS_DEFAULT_BAT_TEST_PWM_RATIO=0
CONFIG_POWER_DISTRIBUTION_QUADROTOR=y
CONFIG_ENABLE_THRUST_BAT_COMPENSATED=y
# CONFIG_LOG_MOTOR_CAP_WARNING is not set

#
# Parameter subsystem
#
# CONFIG_PARAM_SILENT_UPDATES is not set
</pre>

What do you see when you do python3 -m serial.tools.list_ports -v ?
`/dev/ttyS0          
    desc: ttyS0
    hwid: PNP0501
/dev/ttyUSB0        
    desc: USB <-> Serial - USB <-> Serial
    hwid: USB VID:PID=0403:6001 LOCATION=2-2.2
2 ports found
`

Which port are you using? You can check if it is added to the right user group for rights with ls -l /dev/ttyUSB...
`crw-rw---- 1 root dialout 188, 0 Apr  9 08:31 /dev/ttyUSB0`

**Reply by knmcguire** [reply]
Nothing really odd there. so config seems to be the same, the usb port seems to be correct as well and the user rights is good as well... The mystery does continue.... The app does seem to work if you are able to see something on the oscilloscope. 

Additional things that I would try if I'm in the same situation just to try to isolate where the problem is:
- try /dev/ttyS0 (although that is a bit weird, ftdi should show as ttyUSB)
-  switch rx an tx (I always make this mistake, + the ground cavble)
- check for line breakage with a multimeter or try a different ftdi cable
- Try a different computer (if available)

I recently tried the same app on a raspberry pi 5 (raspOS bookworm) as well with both direct uart connection and ftdi cable, but I'll try it on ubuntu 22.04 soon as well once I'm back at the office where the hardware is. maybe there is some additional thing that is necessary to get this to work on ubuntu.

**Reply by tedzaremba** [reply, marked as answer]
Mystery solved! We had a faulty FTDI cable @knmcguire

**Reply by knmcguire** [reply]
aha! good find :smile: ! Glad there was a simple explanation to this, if you weren't able to figure it out after this pass I would have been at loss myself of what ign the problem would have been :sweat_smile:

**Comment by dirksavage88** [comment]
We have a difference in our setup where the crazy radio is not connected (we have one on order). Is there logic in the schematic or a voltage rail that is dependent on the radio for, such as the VCOM power rail?

**Reply by knmcguire** [reply]
Hmm the crazyradio setup doesn't really matter right? It should only matter if the crazyflie has crtp over uart2, which only one connection can exist at the time, but in this case it's only a simple array.
